lib/Coffer/Path.hs

```diff
@@ -0,0 +1,195 @@
+  , mkPath
+  , EntryPath(..)
+  , isoEntryPath
+  , mkEntryPath
+  , entryPathName
+  , entryPathParentDir
+  , entryPathSegments
+  , appendEntryName
+  , pathAsEntryPath
+  , entryPathAsPath
+  , replacePathPrefix
+  ) where
+
+import Data.Text (Text)
+import qualified Data.Text as T
+import Fmt (Buildable, build, pretty)
+import Data.Maybe (fromMaybe)
+import Data.Hashable (Hashable)
+import Data.List.NonEmpty (NonEmpty ((:|)))
+import Control.Lens
+import qualified Data.List.NonEmpty as NE
+import Control.Monad ((>=>))
+import qualified Data.List as List
+
+-- $setup
+-- >>> import Fmt (pretty, build)
+-- >>> import Control.Lens
+-- >>> isRight (Right a) = a
+-- >>> unsafeMkPath = isRight . mkPath
+-- >>> unsafeMkEntryPath = isRight . mkEntryPath
+-- >>> unsafeMkPathSegment = isRight . mkPathSegment
+
+newtype PathSegment = UnsafeMkPathSegment { unPathSegment :: Text }
+  deriving stock (Show, Eq)
+  deriving newtype (Buildable, Hashable)
+
+mkPathSegment :: Text -> Either Text PathSegment
+mkPathSegment segment
+  | T.null segment =
+      Left "Path segments must contain at least 1 character"
+  | T.any (`notElem` pathSegmentAllowedCharacters) segment =
+      Left $ "Path segments can only contain the following characters: '" <> T.pack pathSegmentAllowedCharacters <> "'"
+  | otherwise = Right $ UnsafeMkPathSegment segment
+
+pathSegmentAllowedCharacters :: [Char]
+pathSegmentAllowedCharacters = ['a'..'z'] ++ ['A'..'Z'] ++ ['0'..'9'] ++ "-_"
+
+-- | Path to a directory or an entry.
+newtype Path = Path { unPath :: [PathSegment] }
+  deriving stock (Show, Eq)
+  deriving newtype (Semigroup, Monoid)
+
+makeLensesFor [("unPath", "isoPath")] ''Path
+
+-- |
+-- >>> pretty @Path @String <$> mkPath "a/b/c"
+-- Right "/a/b/c"
+instance Buildable Path where
+  build (Path segments) = "/" <> build (T.intercalate "/" $ pretty <$> segments)
+
+-- | Retrieve a list of a path's segments.
+--
+-- >>> pathSegments $ unsafeMkPath "/a/b/c"
+-- ["a","b","c"]
+pathSegments :: Path -> [Text]
+pathSegments p = p ^.. isoPath . each . to unPathSegment
+
+-- | Parses a path to a directory or entry.
+--
+-- Leading and trailing slashes are optional.
+--
+-- >>> pretty <$> mkPath "/"
+-- Right "/"
+-- >>> pretty <$> mkPath ""
+-- Right "/"
+-- >>> pretty <$> mkPath "/a/"
+-- Right "/a"
+-- >>> pretty <$> mkPath "a"
+-- Right "/a"
+-- >>> pretty <$> mkPath "/a/b/c"
+-- Right "/a/b/c"
+mkPath :: Text -> Either Text Path
+mkPath path = do
+  let segments = path
+        & stripPrefix
+        & stripSuffix
+        & T.splitOn "/"
+
+  if segments == [""]
+    then Right $ Path []
+    else Path <$> traverse mkPathSegment segments
+
+  where
+    stripPrefix s = fromMaybe s $ T.stripPrefix "/" s
+    stripSuffix s = fromMaybe s $ T.stripSuffix "/" s
+
+-- | An entry's full path (directory + entry's name).
+newtype EntryPath = EntryPath { unEntryPath :: NonEmpty PathSegment }
+  deriving stock (Show, Eq)
+
+makeLensesFor [("unEntryPath", "isoEntryPath")] ''EntryPath
```

Comment:
Why `isoEntryPath` and not ~[coerced](https://hackage.haskell.org/package/lens-5.1/docs/Control-Lens-Iso.html#v:coerced)~ `pathSegments`? The `EntryPath` and its segments are entities from domain.

The code style demands it, but I think it would be more readable at use sites as `pathSegments`.

Reply:
I did it this way mostly because there is no function in `Control.Lens.TH` that (1) lets me choose the generated fields' names _and_ (2) generates a `HasPathSegments` typeclass.

`makeLensesWith` does (2) but not (1). `makeLensesFor` does (1) but not (2)  :disappointed: 

Anyway, I just pushed a fixup where I simply wrote my own `HasPathSegments` typeclass ^^